test/integration/conftest.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+from __future__ import annotations
+
+import logging
+import pathlib
+import subprocess
+from typing import Callable, Generator
+
+import jubilant
+import minio
+import pytest
+
+
+@pytest.fixture
+def juju() -> Generator[jubilant.Juju, None, None]:
+    """Make a Juju model with the tracing part of COS ready."""
+    with jubilant.temp_model() as juju:
+        deploy_tempo(juju)
+        deploy_tempo_worker(juju)
+        juju.deploy('minio', config={'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'})
+        juju.deploy('s3-integrator')
+
+        juju.integrate('tempo:s3', 's3-integrator')
+        juju.integrate('tempo:tempo-cluster', 'tempo-worker')
+
+        juju.wait(
+            lambda status: jubilant.all_active(status, 'minio')
+            and jubilant.all_blocked(status, 's3-integrator')
+        )
+
+        address = juju.status().apps['minio'].address
+        mc_client = minio.Minio(
+            f'{address}:9000',
+            access_key='accesskey',
+            secret_key='mysoverysecretkey',
+            secure=False,
+        )
+
+        found = mc_client.bucket_exists('tempo')
+        if not found:
+            mc_client.make_bucket('tempo')
+
+        juju.config('s3-integrator', dict(endpoint=f'http://{address}:9000', bucket='tempo'))
+        juju.run(
+            's3-integrator/0',
+            'sync-s3-credentials',
+            {'access-key': 'accesskey', 'secret-key': 'mysoverysecretkey'},
+        )
+
+        # Tempo goes through a cycle of:
+        # - update own stateful set
+        # - kill own pod
+        # - new pod is scheduled by k8s
+        # - check own stateful set
+        #
+        # This process may take a while.
+
+        juju.wait(jubilant.all_active)
+
+        yield juju
+
+        print(juju.debug_log())
+
+
+@pytest.fixture(scope='session')
+def charm_dir(pytestconfig: pytest.Config) -> Generator[pathlib.Path, None, None]:
```

Comment:
I think future annotations lets you use the shorter return type if you prefer. Fine with me either way.

Reply:
Let me check that, you may be right. It's a 3.9+ feature, but future annotations stringify the type hints and these are not used at run time, so it should work.

Reply:
done, thank you!